### ISSUE TITLE
Omnipose oversegmentic

### ISSUE THREAD

**Issue by NicolaPellicciotta** [MouseLand/cellpose#445]
Dear cellpose/omnipose team,
I am trying to use Cellpose/Omnipose to segment E.coli growing in 2D chamber made of PDMS. 
I am having major issues using Omnipose because it looks like is always over segmenting the cells, despite the flow mask seems to well define the shapes of bacteria. 
It is few days now that i am trying changing parameters (like diameter and thresholds) but results hold the same. It is like the omni algorithm to track filamentous bacteria is not working for me. Any suggestion what could be wrong? 

In attachment few images showing the results from the segmentation of two sample images of my dataset when i used diameter=40 px. The results are always bad also with smaller or larger diameters (or None). I am very surprise because the flow masks seem to recapitulate well the shapes, but the segmentation is bad. In attachment also a picture of the paramters that i use when running cellpose/omnipose eval function. I did it using the colab code available on GitHub (a code-based notebook GPU). 

[Here](https://drive.google.com/file/d/1VE5J_iIwvL2d9x-viEEE7wIHi_rWlI7V/view?usp=sharing) a copy of the colab notebook that i used. It  loads two images  from my google drive (if you want to give it a try on colab).

<img width="1103" alt="omnipose code" src="https://user-images.githubusercontent.com/44974591/154819660-0ba6a276-ad43-445e-aa8f-31876da701a0.png">

<img width="1058" alt="omnipose results" src="https://user-images.githubusercontent.com/44974591/154819655-af2824e2-8621-44fd-ad53-50f47bdd259e.png">

**Comment by guiwitz** [comment]
I ran into the same problem with similar data. After digging a while I found what I believe are two bugs related to omnipose: a missing function ```step_factor``` in the dynamics module and a missing import of omnipose in the model module. See this [image.sc post](https://forum.image.sc/t/omnipose-over-segmenting-filamentous-bacteria/63560/2?) for more details. I fixed these bugs in [this fork](https://github.com/guiwitz/cellpose) and I'm happy to make a PR but it would be nice to have the input from @carsen-stringer as I'm still sceptical myself whether these are actual bugs since they should make omnipose completely unfunctional at the moment.

**Comment by guiwitz** [comment]
Also, now that I can actually use omnipose, there are some weird print outputs that seem to remain from some debugging. I don't know where in the code however. I'm also tagging @kevinjohncutler since this is all very specific to omnipose.
<img width="654" alt="Screenshot 2022-02-21 at 15 53 13" src="https://user-images.githubusercontent.com/4622767/154978863-aeaa716c-d607-494e-9f4c-ed1a88dfe9ae.png">

**Comment by NicolaPellicciotta** [comment]
Hi Guillaume,

thank you so much, this sorted my problem! I am so happy because i was really giving up, and now i see that omnipose algorithm works really well with my images.

I have to admit that i agree with you, in the current state omnipose looks very unstable.
I share the problems that i have found, hoping is going to help improving the future versions:

1- Installing cellpose (pip install cellpose) and then installing omnipose (pip install omnipose). This was giving me a problem such as ‘step_factor not defined’ when i tried to use omnipose.

2- installing cellpose with pip install cellpose[all], (without pip install omnipose) was apparently working but, as Guillaume found out, was not importing omnipose model at all. This resulted in the over segmentation that i have reported.

It would help to have a unique way to install cellpose-omnipose (either with pip install cellpose[all] or with pip install omnipose).
Thanks again

**Comment by kevinjohncutler** [comment]
@NicolaPellicciotta @guiwitz, thanks so much for your feedback and sorry for the trouble. There's been some code refactoring since the original publication of Omnipose, and this is the source of the step factor bug (and those print statements are indeed some debugging lines that I forgot to remove). @guiwitz, please do submit a pull request with your fixes to Cellpose (I must made the same fix in my local version among numerous other edits, as you're right that it makes Omnipose nonfunctional).

I'll clean up those print statements from Omnipose as well and make a new release this week. I will also incorporate some Omnipose-specific tests into the automatic Cellpose tests so things like this don't slip by us in the future (I should have done it a lot earlier!).

**Comment by NicolaPellicciotta** [comment]
Thank you @kevinjohncutler for taking care of this. The omnipose method is very good and i am actually looking forward to use it with all its potential. 

 I wanted just to add that it would be great to have a colab notebook where you show how to use omnipose with some filamentous bacteria/worms images. There are few parameters in the eval function (like skel and cluster) that are not very much documented, and, it would be nice to know how to use them. Also, the dist_threshold value is not clear if is used for the distance flow field of omnipose or  for something related to the cell-probs.

Moreover, it would be great  if you could add also a colab notebook for the training of the model. 
The colab notebooks that are now available (both the test notebook and the training) are for cellpose standard flow field and omnipose is never used. 

Thanks again and sorry for all these requests, but i am sure that researchers like me (that i am not a image scientist) are going to appreciate a lot.

**Comment by guiwitz** [comment]
@kevinjohncutler I made a PR with the fix. Maintaining software is hard, especially when it's not your main job, so I'm happy we could help with this. Omnipose is indeed a great addition to cellpose!

**Comment by kevinjohncutler** [comment]
@guiwitz Agreed, and thank you! 

@NicolaPellicciotta, very good points. I'm hoping to make notebooks and much better documentation soon. Someone else already even made some good colab notebooks that we should link to in the README. Don't be sorry for those requests - I'd want them too! In my opinion, they are a necessary component for any software release, and we will definitely have them done within the next month as we finalize our paper on Omnipose.

**Comment by m-albert** [comment]
_(also posted on https://forum.image.sc/t/omnipose-over-segmenting-filamentous-bacteria)_

Using the cellpose fork pip install git+https://github.com/guiwitz/cellpose.git also worked well for me!

One thing to add from my experience is that scikit-learn needs to be installed for the omnipose import to work. Otherwise cellpose will silently fall back to the cellpose mask reconstruction, which interestingly works well-ish already on bacteria in combination with the pretrained ‘bact_omni’ model.

Thank you @kevinjohncutler for the great omnipose, @NicolaPellicciotta for raising this issue and @guiwitz for sharing your fix!
